File: src/leetCode/java/Solution150.java
package leetCode.java;

import java.util.*;

//Stack O(n) O(num), num為tokens中為數字者
class Solution150 {
    public int evalRPN(String[] tokens) {
        Deque<Integer> deque = new LinkedList<>();
        for(String token : tokens){
            if("+".equals(token)){
                deque.push(deque.poll() + deque.poll());
            }else if("-".equals(token)){
                int a = deque.poll();
                int b = deque.poll();
                deque.push(b - a);
            }else if("*".equals(token)){
                deque.push(deque.poll() * deque.poll());
            }else if("/".equals(token)){
                int a = deque.poll();
                int b = deque.poll();
                deque.push(b / a);
            }else{
                deque.push(Integer.parseInt(token));
            }
        }
        return deque.peek();
    }
}
/* tokens = ["4","13","5","/","-"]
=> 4 - (13 / 5)
*/

File: src/leetCode/c++/Solution150.cpp
//Stack O(n) O(num), n = tokens.size(), num = tokens中為數字者
class Solution150 {
public:
    int evalRPN(vector<string>& tokens) {
        stack<long> st;
        for(string token : tokens){
            //token可能是十位數以上, 不要用char判斷
            if(token != "+" && token != "-" && token != "*" && token != "/"){
                st.push(stoi(token));
            }else{
                long a = st.top(); st.pop();
                long b = st.top(); st.pop();
                if(token == "+") st.push(b + a);
                if(token == "-") st.push(b - a);
                if(token == "*") st.push(b * a);
                if(token == "/") st.push(b / a);
            }
        }
        return st.top();
    }
};